src/main/scala/org/broadinstitute/dsde/firecloud/webservice/OauthApiService.scala

```diff
@@ -0,0 +1,84 @@
+package org.broadinstitute.dsde.firecloud.webservice
+
+import org.broadinstitute.dsde.firecloud.FireCloudConfig
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol.impCurator
+import org.broadinstitute.dsde.firecloud.model.{Curator, UserInfo}
+import org.broadinstitute.dsde.firecloud.service.{FireCloudDirectives, FireCloudRequestBuilding, LibraryService, OAuthService}
+import org.broadinstitute.dsde.firecloud.utils.StandardUserInfoDirectives
+import spray.client.pipelining._
+import spray.http.StatusCodes
+import spray.http.StatusCodes._
+import spray.httpx.SprayJsonSupport._
+import spray.json.DefaultJsonProtocol._
+import spray.json.JsonParser.ParsingException
+import spray.json._
+import spray.routing._
+
+import scala.concurrent.ExecutionContext
+
+trait OauthApiService extends HttpService with FireCloudRequestBuilding
+  with FireCloudDirectives with StandardUserInfoDirectives {
+
+  private implicit val ec: ExecutionContext = actorRefFactory.dispatcher
+
+  val oauthServiceConstructor: () => OAuthService
+
+  def completeWithMissingKey(rc: RequestContext, key: String) = rc.complete(
+    StatusCodes.BadRequest,
+    Map("error" -> Map("summary" -> s"Missing required key '$key'"))
+  )
+
+  def completeWithBadValue(rc: RequestContext, key: String, message: String) = rc.complete(
+    StatusCodes.BadRequest,
+    Map("error" -> Map("summary" -> s"Invalid value for '$key'", "detail" -> message))
+  )
+
+  val oauthRoutes: Route =
+    path("handle-oauth-code") {
+      post { requestContext =>
+        try {
+          val params = requestContext.request.entity.data.asString.parseJson.asJsObject
```

Comment:
I don't care about supporting curl users for login routes, so let's not jump through hoops there.